File: infer/tests/codetoanalyze/java/pulse/taint/UnknownCode.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package codetoanalyze.java.pulse;

import android.content.Intent;
import android.os.Parcel;

class SensitiveClass {
  native Object getId();
}

/** testing how the analysis handles missing/unknown code */
public abstract class UnknownCode {

  static native Object nativeMethod(Object o);

  abstract Object abstractMethod(Object o);

  static void inferSensitiveSink(Object o) {}

  static interface Interface {
    Object interfaceMethod(Object o);
  }

  static void propagateViaUnknownConstructorBad() {
    String source = (String) InferTaint.inferSecretSource();
    // we don't analyze the code for the core Java libraries, so this constructor will be unknown
    String unknownConstructor = new String(source);
    InferTaint.inferSensitiveSink(unknownConstructor);
  }

  static void propagateViaUnknownConstructorOk() {
    String unknownConstructor = new String("");
    InferTaint.inferSensitiveSink(unknownConstructor);
  }

  void propagateViaUnknownCodeOk(Interface i) {
    Object notASource = new Object();
    Object launderedSource1 = nativeMethod(notASource);
    Object launderedSource2 = abstractMethod(launderedSource1);
    Object launderedSource3 = i.interfaceMethod(launderedSource2);
    InferTaint.inferSensitiveSink(launderedSource3);
  }

  void callUnknownSetterBad(Intent i) {
    Object source = InferTaint.inferSecretSource();
    // we don't analyze the source code for Android, so this will be unknown
    i.writeToParcel((Parcel) source, 0);
    InferTaint.inferSensitiveSink(i);
  }

  void propagateEmptyBad() {
    String source = (String) InferTaint.inferSecretSource();
    StringBuffer buffer = new StringBuffer();
    buffer.append(source); // buffer is now tainted
    // even though "" is not tainted, buffer and alias should still be tainted
    StringBuffer alias = buffer.append("");
    InferTaint.inferSensitiveSink(buffer); // should report
    InferTaint.inferSensitiveSink(alias); // should report
  }

  void propagateFootprint(String param) {
    StringBuffer buffer = new StringBuffer();
    buffer.append(param);
    InferTaint.inferSensitiveSink(buffer);
  }

  void callPropagateFootprintBad() {
    propagateFootprint((String) InferTaint.inferSecretSource());
  }

  static native Object nativeMethod2(Object o, int i);

  String propagateTaint(String param) {
    String a = param;
    int i = 1234;
    return (String) nativeMethod2(a, i);
  }

  void callPropagateThenPropagateFootprintBad() {
    String source = (String) InferTaint.inferSecretSource();
    propagateFootprint(propagateTaint(source));
  }

  static void propagateViaInterfaceCodeBad(Interface i) {
    Object source = InferTaint.inferSecretSource();
    Object launderedSource = i.interfaceMethod(source);
    InferTaint.inferSensitiveSink(launderedSource);
  }

  void propagateViaUnknownNativeCodeBad() {
    Object source = InferTaint.inferSecretSource();
    Object launderedSource = nativeMethod(source);
    InferTaint.inferSensitiveSink(launderedSource);
  }

  static void propagateViaUnknownAbstractCodeBad() {
    Object source = InferTaint.inferSecretSource();
    Object launderedSource = nativeMethod(source);
    InferTaint.inferSensitiveSink(launderedSource);
  }

  static void propagatedTaintUnrelated(SensitiveClass sc) {
    Object t1 = nativeMethod(sc.getId());
    Object t2 = nativeMethod(null);
    if (t1 == t2) {
      InferTaint.inferSensitiveSink(t2);
    }
  }

  static String returnToString(Object param) {
    return param.toString();
  }

  static void sourceIndirectToStringToSink() {
    String tainted = returnToString(InferTaint.inferSecretSource());
    InferTaint.inferSensitiveSink(tainted);
  }
}

File: infer/tests/codetoanalyze/kotlin/pulse/taint/UnknownCode.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package codetoanalyze.kotlin.pulse

/** testing how the analysis handles missing/unknown code */
abstract class UnknownCode {

  abstract fun abstractMethod(o: Any?): Any

  interface Interface {
    fun interfaceMethod(o: Any?): Any
  }

  fun propagateViaUnknownCodeOk(i: Interface) {
    val notASource = Any()
    val launderedSource1 = nativeMethod(notASource)
    val launderedSource2 = abstractMethod(launderedSource1)
    val launderedSource3 = i.interfaceMethod(launderedSource2)
    InferTaint.inferSensitiveSink(launderedSource3)
  }

  fun propagateEmptyBad() {
    val source = InferTaint.inferSecretSource() as String
    val buffer = StringBuffer()
    buffer.append(source) // buffer is now tainted
    // even though "" is not tainted, buffer and alias should still be tainted
    val alias = buffer.append("")
    InferTaint.inferSensitiveSink(buffer) // should report
    InferTaint.inferSensitiveSink(alias) // should report
  }

  fun propagateFootprint(param: String?) {
    val buffer = StringBuffer()
    buffer.append(param)
    InferTaint.inferSensitiveSink(buffer)
  }

  fun callPropagateFootprintBad() {
    propagateFootprint(InferTaint.inferSecretSource() as String)
  }

  fun propagateTaint(param: String): String? {
    val i = 1234
    return nativeMethod2(param, i) as String?
  }

  fun callPropagateThenPropagateFootprintBad() {
    val source = InferTaint.inferSecretSource() as String
    propagateFootprint(propagateTaint(source))
  }

  fun propagateViaUnknownNativeCodeBad() {
    val source = InferTaint.inferSecretSource()
    val launderedSource = nativeMethod(source)
    InferTaint.inferSensitiveSink(launderedSource)
  }

  companion object {

    external fun nativeMethod(o: Any?): Any

    fun propagateViaUnknownConstructorBad() {
      val source = InferTaint.inferSecretSource() as String
      // we don't analyze the code for the core Java libraries, so this constructor will be unknown
      val unknownConstructor = java.lang.String(source) as String
      InferTaint.inferSensitiveSink(unknownConstructor)
    }

    fun propagateViaUnknownConstructorOk() {
      val unknownConstructor = ""
      InferTaint.inferSensitiveSink(unknownConstructor)
    }

    external fun nativeMethod2(o: Any?, i: Int): Any?

    fun propagateViaInterfaceCodeBad(i: Interface) {
      val source = InferTaint.inferSecretSource()
      val launderedSource = i.interfaceMethod(source)
      InferTaint.inferSensitiveSink(launderedSource)
    }

    fun propagateViaUnknownAbstractCodeBad() {
      val source = InferTaint.inferSecretSource()
      val launderedSource = nativeMethod(source)
      InferTaint.inferSensitiveSink(launderedSource)
    }
  }
}
